pageserver/src/object_repository.rs

```diff
@@ -157,6 +156,18 @@ impl Repository for ObjectRepository {
             &ObjectValue::ser(&val)?,
         )?;
 
+        // Copy non-rel objects
+        for tag in src_timeline.list_nonrels(at_lsn)? {
+            match tag {
+                ObjectTag::TimelineMetadataTag => {} // skip it
+                _ => {
+                    let img = src_timeline.get_page_at_lsn_nowait(tag, at_lsn)?;
+                    let val = ObjectValue::Page(PageEntry::Page(img));
+                    let key = ObjectKey { timeline: dst, tag };
+                    self.obj_store.put(&key, at_lsn, &ObjectValue::ser(&val)?)?;
+                }
+            }
+        }
```

Comment:
I decided to copy nonrel data to new time line because
1. Size of this data is expected too be small enough, so copying them should not cause large space overhead
2. Simplify copying this data to tarball prepared for launch of computation node.
Nonrel data is needed only for preparing this tarball. So optimizing access to it may be not so critical. From the other side,
node startup time may be one of the most critical characteristics of zenith.

In any case, I think there are many questions to branches themselves: how we are going to use and maintain them? Do we need branches at all...  Right now presence if branches cause many problems with GC. And it is not clear how them should be integrated with new storage snapshots. May be creation of the branch requires making snapshot. If so, copying of nonrel data to new timeline may be considered as step in this direction.

Reply:
Thanks. I see the reason now.

> Size of this data is expected too be small enough, so copying them should not cause large space overhead

But if we don't copy them, there's no space overhead at all. And the overhead of iterating back to the ancestor timeline is negligible, isn't it?

I prefer to consider non-rel objects as first-class citizens in our object store and expect that they will follow the same APIs where possible, diverging from them only when it's inevitable.

>In any case, I think there are many questions to branches themselves: how we are going to use and maintain them? Do we need branches at all...

That's a good question. I was thinking of branches as one of our major features. Let's discuss it at today's meeting.

>And it is not clear how them should be integrated with new storage snapshots.

Could you explain your concerns? I don't see any issues with it. In my understanding, when building a snapshot, we simply collect a full (or incremental) history of each object.  If necessary, we'll iterate back to the ancestor's timeline and get some data from it.

Reply:
> But if we don't copy them, there's no space overhead at all. And the overhead of iterating back to the ancestor timeline is negligible, isn't it?
Not sure. I have not measured, but all objects from one timeline can be extracted using one lookup in object storage and iteration through results. And extracting objects from N timelines will require at least N lookups and may be even N*K where K is total of non-rel objects. The difference between 1 and N*K can be very noticable.

Reply:
> Could you explain your concerns? I don't see any issues with it. In my understanding, when building a snapshot, we simply collect a full (or incremental) history of each object. If necessary, we'll iterate back to the ancestor's timeline and get some data from it.

There is a lot of misunderstanding and terms mixing here. What do we mean by "snapshot", how and when it is produced...
Based on new storage model proposed by Eric, I have something like this in my mind:
1. We have in-memory KV storage where we collect received WAL records.
2. Periodically  we apply sequences of this WAL records and materialize most recent version of the page. Materialized pages are stored in some persistent storage. Periodically we make incremental snapshots of this persistent storage and upload it to S3.
3. User may request explicit snapshot. We can call it a "branch", but IMHO it is confusing term. In most cases users need to specify recovery point, i.e. snapshot rather than create alternative timeline.
4. Branches or forks may also exists, but it seems to be more or less experimental feature needed mostly for debugging and performing various experiments. As far as branches can not be merged, branch is always "dead-end".
5. Do we need PITR for branches? Do we need PITR at all?  If the question is "no", then branch and snapshot are very similar. 

IMHO current hierarchy of branches more introducing complexity rather provide some useful functionality.
As far me, I prefer to start with linear snapshot history (as in probackup) and leave branching support for future.